contracts/axelar-gateway/src/test_auth.rs

```diff
@@ -0,0 +1,551 @@
+    env.as_contract(&contract_id, || {
+        assert_err!(
+            auth::validate_proof(&env, &msg_hash, proof),
+            GatewayAuthError::InvalidSigners
+        );
+    })
+}
+
+#[test]
+fn fail_validate_proof_threshold_not_met() {
+    let (env, contract_id, client) = setup_env();
+    let user = Address::generate(&env);
+
+    let signers = initialize(&env, &client, user, randint(0, 10), randint(1, 10));
+
+    let env = &env;
+    let mut total_weight = 0u128;
+
+    let msg_hash = generate_random_payload_and_hash(env);
+    let mut proof = generate_proof(env, msg_hash.clone(), signers);
+
+    // Modify signatures to make them invalid
+    let mut new_signers = Vec::new(env);
+    for ProofSigner { signer, signature } in proof.signers.iter() {
+        total_weight += signer.weight;
+
+        if total_weight < proof.threshold {
+            new_signers.push_back(ProofSigner { signer, signature });
+        } else {
+            new_signers.push_back(ProofSigner {
+                signer,
+                signature: ProofSignature::Unsigned,
+            });
+        }
+    }
+    proof.signers = new_signers;
+
+    // should panic, all signatures are valid but total weight is below threshold
+    env.as_contract(&contract_id, || {
+        assert_err!(
+            auth::validate_proof(&env, &msg_hash, proof),
+            GatewayAuthError::InvalidSignatures
+        );
+    })
+}
+#[test]
+fn fail_validate_proof_threshold_overflow() {
+    let (env, contract_id, client) = setup_env();
+    let user = Address::generate(&env);
+
+    let mut signers = initialize(&env, &client, user, randint(0, 10), randint(1, 10));
+
+    let env = &env;
+
+    let last_index = signers.signers.signers.len() - 1;
+
+    // get last signer and modify its weight to max u128 - 1
+    if let Some(mut last_signer) = signers.signers.signers.get(last_index) {
+        last_signer.weight = u128::MAX - 1;
+        signers.signers.signers.set(last_index, last_signer);
+    }
+
+    let msg_hash = generate_random_payload_and_hash(env);
+    let proof = generate_proof(env, msg_hash.clone(), signers);
+
+    // should panic, as modified signer wouldn't match the epoch
+    env.as_contract(&contract_id, || {
+        assert_err!(
+            auth::validate_proof(&env, &msg_hash, proof),
+            GatewayAuthError::InvalidSigners
+        );
+    });
+}
+
+#[test]
+fn test_rotate_signers() {
+    let (env, contract_id, client) = setup_env();
+
+    let user = Address::generate(&env);
+    let previous_signer_retention = 1;
+
+    let signers = initialize(
+        &env,
+        &client,
+        user.clone(),
+        previous_signer_retention,
+        randint(1, 10),
+    );
+
+    let msg_hash = generate_random_payload_and_hash(&env);
+
+    let new_signers = generate_signers_set(&env, randint(1, 10), signers.domain_separator);
+
+    testutils::rotate_signers(&env, &contract_id, new_signers.clone());
```

Comment:
we should test gateway.rotate_signers directly instead of calling the internal auth module. It's ok if we need to do it for a specific unit test, but I'd rather just test via the contract entrypoints whenever it makes sense. We should probably move most of the tests into the integration `tests` folder in a separate PR to distinguish from unit tests. And unit tests can live directly in the module itself. Is it possible to call `auth` methods in a unit test with a mock env without having to register a contract as done in setup_env?

Reply:
> but I'd rather just test via the contract entrypoints whenever it makes sense

Most of the tests in `test_gateway.rs` call the proper contract entrypoints. In the error cases, they're more focused on catching the errors that are raised before calling `auth::...`. It'd probably be possible to construct the tests such that they cover most of the things that `test_auth` hits, but it's also nice imo that `test_auth` is specifically testing the functions in `auth.rs`

> Is it possible to call auth methods in a unit test with a mock env without having to register a contract as done in setup_env?

I don't think it'd be possible since interfacing with the env without `.as_contract()` brings up this error:
```
this function is not accessible outside of a contract, wrap the call with `env.as_contract()` to access it from a particular contract
```
but it'd be worth looking into

Reply:
As per discussion on the sync, convert these into integration tests. If we're testing specific methods that aren't easily testable via integration tests, we can write unit tests in those specific instances